File: 21-merge-two-sorted-lists/21-merge-two-sorted-lists.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def mergeTwoLists(self, list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:
        if list1 == None and list2 == None:
            return None
        if list1 == None:
            return list2
        if list2 == None:
            return list1
        
        head = ListNode()
        current = head
        
        left = list1
        right = list2
        
        while (left != None and right != None):
            if (left.val < right.val):
                current.val = left.val
                left = left.next
            else:
                current.val = right.val
                right = right.next
            current.next = ListNode()
            current = current.next
            
        while (left != None):
            current.val = left.val
            if (left != None and left.next != None):
                current.next = ListNode()
                current = current.next
            left = left.next
        
        while (right != None):
            current.val = right.val
            if (right != None and right.next != None):
                current.next = ListNode()
                current = current.next
            right = right.next
            
        
        return head

File: 21-merge-two-sorted-lists/21-merge-two-sorted-lists.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode mergeTwoLists(ListNode list1, ListNode list2) {
        ListNode left = list1;
        ListNode right = list2;
        ListNode head = new ListNode();
        ListNode current = head;
        
        while (left != null && right != null) {
            if (left.val < right.val) {
                current.next = left;
                left = left.next;
            } else {
                current.next = right;
                right = right.next;
            }
            current = current.next;
        }
        
        while (left != null) {
            current.next = left;
            left = left.next;
            current = current.next;
        }
        
        while (right != null) {
            current.next = right;
            right = right.next;
            current = current.next;
        }
        
        return head.next;
    }
}